hs628_tools/regression_check.py

```diff
@@ -0,0 +1,189 @@
+        # F-stat,F-statpval,ResidualSumofSquares,RegressionMetric
+        self.colnames = None
+        self.rsquared = 0
+        self.fstatistic = 0
+        self.fstat_pval = 0
+        self.resid_sumsq = 0
+        self.regmetric = 0
+
+    def check_linearity_residual_homoscedasticity_independence_plt(self, X, y, model):
+        """
+        Checks to see if the data can fit to a linear model/non-linear model,checks the homoscedasticity
+         of the residuals/error terms and also checks for their independence.
+            -----------
+            Param:
+                X: 2-D numpy array of the predictors
+                y: 1-D numpy array of the response variables
+                linear : Linear regression model fitted to X(predictors) and y(response)
+            -----------
+            Return:
+                Doesn't return anything, just shows the plot object built
+        """
+
+        # Plot of Residuals vs predicted values
+        fig, axes = plt.subplots(1, 1, sharex=False, sharey=False)
+        fig.suptitle('[Residual Plots]')
+        fig.set_size_inches(12, 5)
+        # plotting the predicted values from the linear model vs the error values
+        axes.plot(model.predict(X), y - model.predict(X), 'bo--')
+        axes.axhline(y=0, color='k')
+        axes.grid()
+        axes.set_title('Linear/Non-Linear')
+        axes.set_xlabel('predicted values')
+        axes.set_ylabel('residuals')
+        plt.show()
+
+    def check_residuals_normality_plt(self, X, y, model):
+        """
+        Checks to see the normality of the error terms through a density plot of the residuals
+            -----------
+            Param:
+                X: 2-D numpy array of the predictors
+                y: 1-D numpy array of the response variables
+                linear : Linear regression model fitted to X(predictors) and y(response)
+            -----------
+            Return:
+                Doesn't return anything, just shows the plot object built
+        """
+        residuals = y - model.predict(X)
+
+        # Density plot of the residuals
+        sns.distplot(residuals)
+        plt.xlabel("Residuals")
+        plt.title('Residual Normality Check')
+        plt.show()
+
+    def check_predictors_independence_plt(self, X, y, colnames):
+        """
+        Checks to see the independence of the predictors by plotting a heatmap of the correlations
+            -----------
+            Param:
+                X: 2-D numpy array of the predictors
+                y: 1-D numpy array of the response variables
+                colnames : Column names of the predictors
+            -----------
+            Return:
+                Doesn't return anything, just shows the plot object built
+        """
+        self.colnames = colnames
+        data_df = pd.DataFrame(X, columns=self.colnames)
+
+        # Heatmap of the correlation matrix
+        sns.heatmap(data_df.corr(), cmap='binary', annot=True)
```

Comment:
Although using pandas like this would certainly be fine down the road, recall that the project assignment has a restriction against using pandas (generally for the purpose of focusing attention on improving familiarity with sklearn and numpy before moving on to pandas).  It's easy enough to calculate this correlation matrix with a numpy method.

Reply:
Yes, I forgot about that. Let me try to redo in numpy. You could just say this was me being lazy :)